src/inventory_ui.cpp

```diff
@@ -770,86 +773,270 @@ void inventory_selector::remove_dropping_items( player &u ) const
+            break;
+        } else if( action == "QUIT" ) {
+            return std::list<std::pair<int, int> >();
         }
     }
-    return display_slice(reduced_inv, title);
+
+    std::list<std::pair<int, int>> dropped_pos_and_qty;
+
+    for( auto drop_pair : dropping ) {
+        int num_to_drop = drop_pair.second;
+        if( num_to_drop == -1 ) {
+            num_to_drop = inventory::num_items_at_position( drop_pair.first );
+        }
+        dropped_pos_and_qty.push_back( std::make_pair( drop_pair.first, num_to_drop ) );
+    }
+
+    return dropped_pos_and_qty;
 }
 
-int game::inv_for_salvage(const std::string &title, const salvage_actor& actor )
+// Display current inventory.
+int game::inv( const int position )
 {
-    u.inv.restack(&u);
+    u.inv.restack( &u );
+    u.inv.sort();
+
+    return inventory_selector( u ).execute_pick( _( "Inventory:" ), position );
+}
+
+int game::inv_for_filter( const std::string &title, item_filter filter, const std::string &none_message )
+{
+    u.inv.restack( &u );
     u.inv.sort();
-    indexed_invslice reduced_inv = u.inv.slice_filter_by_salvageability( actor );
-    return display_slice(reduced_inv, title);
+
+    inventory_selector inv_s( u, filter );
+
+    if( inv_s.empty() ) {
+        const std::string msg = ( none_message.empty() ) ? _( "You don't have the necessary item." ) : none_message;
+        popup_getkey( msg.c_str() );
+        return INT_MIN;
+    }
+
+    return inv_s.execute_pick( title );
+}
+
+int game::inv_for_all( const std::string &title, const std::string &none_message )
+{
+    const std::string msg = ( none_message.empty() ) ? _( "Your inventory is empty." ) : none_message;
+    return inv_for_filter( title, allow_all_items, msg );
+}
+
+int game::inv_for_activatables( const player &p, const std::string &title )
+{
+    return inv_for_filter( title, [ &p ]( const item &it ) {
+        return p.rate_action_use( it ) != HINT_CANT;
+    }, _( "You don't have any items you can use." ) );
+}
+
+int game::inv_for_flag( const std::string &flag, const std::string &title )
+{
+    return inv_for_filter( title, [ &flag ]( const item &it ) {
+        return it.has_flag( flag );
+    } );
 }
 
-item_location game::inv_map_splice( item_filter filter, const std::string &title, int radius )
+int game::inv_for_id( const itype_id &id, const std::string &title )
 {
-    return inv_map_splice( filter, filter, filter, title, radius );
+    return inv_for_filter( title, [ &id ]( const item &it ) {
+        return it.type->id == id;
+    }, string_format( _( "You don't have a %s." ), item::nname( id ).c_str() ) );
+}
+
+int game::inv_for_tools_powered_by( const itype_id &battery_id, const std::string &title )
+{
+    return inv_for_filter( title, [ &battery_id ]( const item & it ) {
+        return it.is_tool() && it.ammo_type() == battery_id;
+    }, string_format( _( "You don't have %s-powered tools." ), item::nname( battery_id ).c_str() ) );
```

Comment:
The function should accept `ammotype` instead of `itype_id` as the `battery_id`. And use `ammo_name` here to get the name of the fuel type.